users/cacher/cacher.go

```
@@ -0,0 +1,188 @@
+	"go.uber.org/zap"
+
+	"github.com/KyberNetwork/reserve-stats/lib/influxdb"
+	logSchema "github.com/KyberNetwork/reserve-stats/tradelogs/storage/schema/tradelog"
+	"github.com/KyberNetwork/reserve-stats/users/common"
+	"github.com/KyberNetwork/reserve-stats/users/storage"
+)
+
+const (
+	influxDB    = "trade_logs"
+	expireTime  = time.Hour
+	richPrefix  = "rich"
+	kycedPrefix = "kyced"
+)
+
+//RedisCacher is instance for redis cache
+type RedisCacher struct {
+	sugar          *zap.SugaredLogger
+	postgresDB     *storage.UserDB
+	influxDBClient client.Client
+	redisClient    *redis.Client
+	kycedCap       *common.UserCap
+	nonKycedCap    *common.UserCap
+}
+
+//NewRedisCacher returns a new redis cacher instance
+func NewRedisCacher(sugar *zap.SugaredLogger, postgresDB *storage.UserDB,
+	influxDBClient client.Client, redisClient *redis.Client) *RedisCacher {
+	return &RedisCacher{
+		sugar:          sugar,
+		postgresDB:     postgresDB,
+		influxDBClient: influxDBClient,
+		redisClient:    redisClient,
+		kycedCap:       common.NewUserCap(true),
+		nonKycedCap:    common.NewUserCap(false),
+	}
+}
+
+//CacheUserInfo save user info to redis cache
+func (rc *RedisCacher) CacheUserInfo() error {
+	if err := rc.cacheAllKycedUsers(); err != nil {
+		return err
+	}
+	if err := rc.cacheRichUser(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (rc *RedisCacher) cacheAllKycedUsers() error {
+	var (
+		logger    = rc.sugar.With("func", "user/cacher/cacheAllKycedUsers")
+		addresses []string
+		err       error
+	)
+	// read all address from addresses table in postgres
+	if addresses, err = rc.postgresDB.GetAllAddresses(); err != nil {
+		logger.Errorw("error from query postgres db", "error", err.Error())
+		return err
+	}
+	logger.Debugw("addresses from postgres", "addresses", addresses)
+
+	pipe := rc.redisClient.Pipeline()
+	defer func() {
+		if _, err := pipe.Exec(); err != nil {
+			logger.Errorw("pipeline exec error", "error", err.Error())
+		}
+	}()
```

the error is ignored, `pipe.Exec` failed. So `pipe.Discard/Close` if `saveToCache` failed.